### PR TITLE
feat: add tree view for offices component

### PR DESCRIPTION
**Pull request by Rishabh106**
## Description
Added Tree View in Offices Component and designed its UI.

Features which deviate from community app:
- Added View Office to show on clicking office in the tree.
- Added 'Close Button' in View Office to close the current view for office.

## Related issues and discussion
#149 

## Screenshots, if any
tree_view_3
tree_view_2

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

File: src/app/organization/offices/office-tree.service.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  /**
+   * Builds and returns the offices tree.
+   * @param {any} officeData Offices data.
+   * @returns {OfficeNode} Offices tree nodes.
+   */
+  buildOfficeTree(officesData: any): OfficeNode[] {
+    const officeTree: OfficeNode[] = [];
+
+    // Header nodes
+    officeTree.push(new OfficeNode(officesData[0].name, officesData[0].id));
```
Comment (punwai) [inline src/app/organization/offices/office-tree.service.ts]: I know that I have suggested you to do this myself, but lets make sure that the thing that we push first is the Main Head Office, incase the array comes sorted so that the Main Head Office is not the first element.
```suggestion
    const mainOffice = officesData.find((office: any) => office.hierarchy === '.');
    officeTree.push(new OfficeNode(mainOffice.name, mainOffice.id));
```

File: src/app/organization/offices/office-tree.service.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    // Add offices to any array where index for each is denoted by its id
+    for (const office of officesData) {
+      offices[office.id] =
+        new OfficeNode(office.name, office.id, office.parentId);
+    }
+
+    // Construct offices tree by adding all nodes belonging to headers (with parent id = 0),
+    // and rest as children to respective parent nodes.
+    for (const office of officesData) {
+      if (office.id !== 1) {
```
Comment (punwai) [inline src/app/organization/offices/office-tree.service.ts]: Again to ensure that we are dealing with the parent component
```suggestion
      if (office.hierarchy !== '.') {
```

File: src/app/organization/offices/office-tree.service.ts (hunk 1 above)
Comment (punwai) [reply src/app/organization/offices/office-tree.service.ts]: I am sorry if this does not run, I am quite busy at the moment so I couldn't personally test this myself.

Comment (Rishabh106) [reply src/app/organization/offices/office-tree.service.ts]: It is working, no errors present. Thanks for the suggestion :)

File: src/app/organization/offices/offices.component.html [hunk 3]
```diff
@@ -60,3 +66,47 @@
   </div>
 
 </div>
+
+<!-- Tree View -->
+
+<div class="container" fxLayout="column" [hidden]="viewGroup.value !== 'treeView'">
```
Comment (abhaychawla) [inline src/app/organization/offices/offices.component.html]:
```suggestion
<div class="container" [hidden]="viewGroup.value !== 'treeView'">
```

**Comment by Rishabh106** [comment]
> Great work!! I have reviewed your code, I just want to run it on my machine and then I will approve the task, I will do it by tonight as I am currently in Office. I know the deadline of submission is tonight but don't worry about it. The deadline for Mentor to review is till tomorrow.
> 
> Also, similar to Chart Account were some information is shown for the account when it is clicked. Could you check the possibility of adding the same for an Office? This would be an additional enhancement which you can work on separately (not as a part of the task).

I will try to do that, how much information about the office needs to be displayed?

**Comment by abhaychawla** [comment]
We can try something similar to what is displayed while viewing an office (if possible)

**Comment by Rishabh106** [comment]
> We can try something similar to what is displayed while viewing an office (if possible)

I'm thinking of adding

- Name of Office
- Parent Office (with ngIf directive)
- Opened On
- External ID (with ngIf directive)

**Comment by Rishabh106** [comment]
> We can try something similar to what is displayed while viewing an office (if possible)

Screenshot_5
Is this layout fine?
I've added getDatatables method in Organization Service so it can be used in the future for other components too. I'll push the changes once I get a green light from you.

**Comment by punwai** [comment]
Could I recommend you align expand all and collapse all to the left, or shift it so it is in line with the trees card when the information card appears.

**Comment by punwai** [comment]
Apart from this, layout looks very good

**Comment by Rishabh106** [comment]
tree_view_2
Moved it all to the left for now.

**Comment by punwai** [comment]
Also, it is nothing major, but try to ensure the shadows and everything are consistent in the component. I can see that the left tree has a much stronger shadow and a sharp corner. If decide use a card for the information, it would be better to do the same for the tree.

**Comment by Rishabh106** [comment]
> Also, it is nothing major, but try to ensure the shadows and everything are consistent in the component. I can see that the left tree has a much stronger shadow and a sharp corner. If you decide to use a card for the information, it would be better to do the same for the tree.

Keeping the mat-tree in a card doesn't affect its shadows so it is not possible to do that.

File: src/app/organization/offices/offices.component.html [hunk 4]
```diff
@@ -60,3 +66,119 @@
...
+      Expand All
+    </button>
+    <button mat-raised-button (click)="nestedTreeControl.collapseAll()">
+      Collapse All
+    </button>
+  </div>
+
+  <div fxLayout="row" fxLayoutGap="4%" fxLayout.lt-md="column">
+
+    <div class="mat-elevation-z8" fxFlex>
```
Comment (punwai) [inline src/app/organization/offices/offices.component.html]: I meant this. The mat-elevation-z8 creates a stronger shadow than the card actually does, and it also does not provide the same edge corners as the card, which I think looks just a tiny bit inconsistent. Again, nothing major, but I just think it might look abit nicer.

Comment (Rishabh106) [reply src/app/organization/offices/offices.component.html]: Thanks for the suggestion, I did not realize that.

File: src/app/organization/offices/offices.component.html [hunk 5]
```diff
@@ -60,3 +66,127 @@
...
+
+<div class="container" [hidden]="viewGroup.value !== 'treeView'">
+
+  <div class="m-b-20" fxLayout="row" fxLayoutAlign="start" fxLayoutGap="20px">
+    <button mat-raised-button (click)="nestedTreeControl.expandAll()">  <!-- Bug(angular-material): https://github.com/angular/material2/issues/12170 -->
+      Expand All
+    </button>
+    <button mat-raised-button (click)="nestedTreeControl.collapseAll()">
+      Collapse All
+    </button>
```
Comment (muskankhedia) [inline src/app/organization/offices/offices.component.html:80]: @Rishabh106 Everything looks great and fine. Please look over the comment in the [link](https://github.com/angular/components/issues/12170#issuecomment-406530386). As if you go through the test cases written for that particular component of nestedTreeControl, you can see `treeControl.datanodes` have been provided with a value which helps for collapseAll() and expandAll().  Give it a try if it works. Let me know regarding teh updates.

Comment (Rishabh106) [reply src/app/organization/offices/offices.component.html:80]: In which file do I need to add the dataNodes value? (Sorry if its too obvious but I'm still a beginner at Angular)

Comment (muskankhedia) [reply src/app/organization/offices/offices.component.html:80]: Yaa sure no problem. Feel free to ask any doubt.
You need to add another key `datanodes` for the `treeControl` in the `office.component.ts` file.

Comment (Rishabh106) [reply src/app/organization/offices/offices.component.html:80]: I got it working, should I push the changes?

Comment (muskankhedia) [reply src/app/organization/offices/offices.component.html:80]: Can you make another commit for that, will review and if any problem we can revert back? If everything goes fine, you can rebase it.

Comment (Rishabh106) [reply src/app/organization/offices/offices.component.html:80]: Done, I'll resolve the merge conflicts after everything else is fine.

Comment (muskankhedia) [reply src/app/organization/offices/offices.component.html:80]: It's working well, you can rebase it.

**Comment by muskankhedia** [comment]
@Rishabh106 Please drop all the commits except yours, to bring it to the initial state. Then you can follow the steps as mentioned [here](https://github.com/openMF/web-app/pull/664#issuecomment-626359832). In case of any problems keep your code safe in another branch locally or keep a copied folder of the code.

**Comment by Rishabh106** [comment]
Sorry, I messed up as there were about 35 commits which had to be pulled. Illustration drop them and then resolve the merge conflicts.

**Comment by Rishabh106** [comment]
> 
> 
> @Rishabh106 Can you resolve MCs please?

Yes, I've resolved them. Please inform me if I messed up something while resolving the MCs, haven't been active lately so there's room for mistake.

File: src/app/organization/offices/offices.component.ts [hunk 6]
```diff
@@ -1,13 +1,19 @@
 /** Angular Imports */
 import { Component, OnInit, ViewChild } from '@angular/core';
-import { MatPaginator } from '@angular/material/paginator';
-import { MatSort } from '@angular/material/sort';
-import { MatTableDataSource } from '@angular/material/table';
```
Comment (karantakalkar) [inline src/app/organization/offices/offices.component.ts:5]: Restore them, imports are more specific in v9

File: src/app/organization/organization-routing.module.ts [hunk 7]
```diff
@@ -64,8 +64,9 @@ import { FundMappingComponent } from './fund-mapping/fund-mapping.component';
 import { CreateHolidayComponent } from './holidays/create-holiday/create-holiday.component';
 
 /** Custom Resolvers */
-import { LoanProvisioningCriteriaResolver } from './loan-provisioning-criteria/common-resolvers/loan-provisioning-criteria.resolver';
-import { OfficesResolver } from './offices/common-resolvers/offices.resolver';
+import { LoanProvisioningCriteriaResolver } from './loan-provisioning-criteria/loan-provisioning-criteria.resolver';
+import { OfficesResolver } from './offices/offices.resolver';
```
Comment (karantakalkar) [inline src/app/organization/organization-routing.module.ts:68]: undo these changes

**Review by karantakalkar — CHANGES_REQUESTED** [review]
@Rishabh106 the app has been upgraded to angular 9, you will need to run `npm install` to update dependencies localy.

File: src/app/organization/organization.service.ts [hunk 8]
```diff
@@ -783,4 +783,12 @@ export class OrganizationService {
     return this.http.post(`${urlSuffix}/uploadtemplate`, formData , { params: httpParams } );
   }
 
+  /*
+   * @param {string} appTableName Application table name.
+   * @returns {Observable<any>} Datatables data.
+   */
+  getDatatables(appTableName: string): Observable<any> {
+    return this.http.get(`/datatables?apptable=${appTableName}`);
```
Comment (muskankhedia) [inline src/app/organization/organization.service.ts:791]:
```suggestion
    const httpParams = new HttpParams().set('apptable', appTableName);
    return this.http.get(`/datatables`, { params: httpParams });
```

**Review by muskankhedia — CHANGES_REQUESTED** [review]

File: src/app/organization/offices/office-data-tables.resolver.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
+/** Angular Imports */
```
Comment (karantakalkar) [inline src/app/organization/offices/office-data-tables.resolver.ts:1]: There is already an offices data tables resolver in common resolvers folder use that.

File: src/app/organization/organization.service.ts [hunk 10]
```diff
@@ -783,4 +783,13 @@ export class OrganizationService {
...
   }
 
+  /*
+   * @param {string} appTableName Application table name.
+   * @returns {Observable<any>} Datatables data.
+   */
+  getDatatables(appTableName: string): Observable<any> {
+    const httpParams = new HttpParams().set('apptable', appTableName);
+    return this.http.get(`/datatables`, { params: httpParams });
+  }
```
Comment (karantakalkar) [inline src/app/organization/organization.service.ts:793]: remove this getOfficeDatatables() function is already there in service

File: src/app/organization/organization-routing.module.ts [hunk 11]
```diff
@@ -66,6 +66,7 @@ import { CreateHolidayComponent } from './holidays/create-holiday/create-holiday
 /** Custom Resolvers */
 import { LoanProvisioningCriteriaResolver } from './loan-provisioning-criteria/common-resolvers/loan-provisioning-criteria.resolver';
 import { OfficesResolver } from './offices/common-resolvers/offices.resolver';
+import { OfficeDataTablesResolver } from './offices/office-data-tables.resolver';
```
Comment (karantakalkar) [inline src/app/organization/organization-routing.module.ts:69]:
```suggestion

```

File: src/app/organization/organization-routing.module.ts [hunk 12]
```diff
@@ -164,7 +165,8 @@ const routes: Routes = [
               path: '',
               component: OfficesComponent,
               resolve: {
-              offices: OfficesResolver
+              offices: OfficesResolver,
+              officeDataTables: OfficeDataTablesResolver
```
Comment (karantakalkar) [inline src/app/organization/organization-routing.module.ts:169]:
```suggestion
              officeDataTables: OfficeDatatablesResolver
```
Resolver already there in common resolvers

File: src/app/organization/organization-routing.module.ts [hunk 13]
```diff
@@ -657,6 +659,7 @@ const routes: Routes = [
   providers: [
     LoanProvisioningCriteriaResolver,
     OfficesResolver,
+    OfficeDataTablesResolver,
```
Comment (karantakalkar) [inline src/app/organization/organization-routing.module.ts:662]:
```suggestion

```

**Review by karantakalkar — CHANGES_REQUESTED** [review]
We can merge after this

**Comment by abhaychawla** [comment]
@karantakalkar is this PR now ready to be merged?

**Comment by bharathcgowda** [comment]
@Rishabh106  @karantakalkar @abhaychawla  @muskankhedia  is this PR ready to be merged ??

**Comment by Rishabh106** [comment]
> @Rishabh106 @karantakalkar @abhaychawla @muskankhedia is this PR ready to be merged ??

should I resolve the merge conflicts?
